src/codegen/riscv64/assembler-riscv64.cc

```diff
@@ -1696,74 +1696,159 @@ void Assembler::sfence_vma(Register rs1, Register rs2) {
 
 void Assembler::nop() { addi(ToRegister(0), ToRegister(0), 0); }
 
-// The code in *this function only* is based on LLVM's `generateInstSeq`
-// (RISCVMatInt.cpp), part of the LLVM Project, under the Apache License v2.0
-// with LLVM Exceptions. See https://llvm.org/LICENSE.txt for license
-// information. SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 void Assembler::RV_li(Register rd, int64_t imm) {
+  // 64-bit imm is put in the register rd.
+  // In most cases the imm is 32 bit and 2 instructions are generated. If a
+  // temporary register is available, in the worst case, 6 instructions are
+  // generated for a full 64-bit immediate. If temporay register is not
+  // available the maximum will be 8 instructions. If imm is more than 32 bits
+  // and a temp register is available, imm is deivided into two 32-bit parts,
+  // low_32 and up_32. Each part is built in a separate register. low_32 is
+  // built before up_32. If low_32 is negative (upper 32 bits are 1), 0xffffffff
+  // is subtracted from up_32 before up_32 is built. This compensates for 32
+  // bits of 1's in the lower when the two registers are added. If no temp is
+  // available, the upper 32 bit is built in rd, and the lower 32 bits are
+  // devided to 3 parts (11, 11, and 10 bits). The parts are shifted and added
+  // to the upper part built in rd.
   if (is_int32(imm + 0x800)) {
-    // Depending on the active bits in the immediate Value v, the following
-    // instruction sequences are emitted:
-    //
-    // v == 0                        : ADDI
-    // v[0,12) != 0 && v[12,32) == 0 : ADDI
-    // v[0,12) == 0 && v[12,32) != 0 : LUI
-    // v[0,32) != 0                  : LUI+ADDI(W)
-    int64_t Hi20 = ((imm + 0x800) >> 12);
-    int64_t Lo12 = imm << 52 >> 52;
-    Register base = zero_reg;
-
-    if (Hi20) {
-      lui(rd, (int32_t)Hi20);
-      base = rd;
-    }
-
-    if (Lo12 || Hi20 == 0) {
-      addi(rd, base, Lo12);
+    // 32-bit case. Maximum of 2 instructions generated
+    int64_t high_20 = ((imm + 0x800) >> 12);
+    int64_t low_12 = imm << 52 >> 52;
+    if (high_20) {
+      lui(rd, (int32_t)high_20);
+      if (low_12) {
+        addi(rd, rd, low_12);
+      }
+    } else {
+      addi(rd, zero_reg, low_12);
     }
     return;
+  } else {
+    // 64-bit case: devid imm to two 32-bit parts, upper and lower
+    if (imm == -1) {
```

Comment:
Why this special case for -1? if `imm` is -1, the `if` above would be true (-1 + 0x800 = 0x7ff which is an int32).